### DISCUSSION TITLE
v0.21.0: 2025/07/01 🌞 Summer Edition 🌞

### DISCUSSION THREAD

**Discussion by janpfeifer** [gomlx/gomlx#189] [Announcements]
* Package `simplego`:
  * Added `GetBackend` that returns a singleton backend, created with the default configuration at the first request.
* Package `ui/commandline`:
  * Added optional extra arbitrary metrics to print in the command-line with `AttachProgressBar`.
  * Added `FormatDuration` to pretty-print duration.
* Package `graph`
  * Added gradients of `Cos` and `Sin` that were missing.
  * Fixed (removed) the extra empty line in auto-generate functions comments that was preventing the documentation
     from being assigned to the functions.
  * Added parameters `sorted` and `unique` to `Scatter` (like the other functions `Scatter*`) -- **Small API change**.
  * Added `ScatterUpdate`, for now only for `unique=true`.
  * Package `nanlogger`:
    * Allow traces that only report also.
    * Created context parameter `optimizer.ParamNanLogger`: if set to NanLogger, it will trace all occurrences of
      of NaN values in gradient: great to debug where are the NaN appearing in the model first.
* Package `ml/train`:
  * Improved support for accumulated gradients. Fixed evaluation (context reuse) for when using accumulated gradients.
  * Added `Trainer.WithMaxExecutors`.
* Package `ml/train/metrics`:
  * `MeanMetric` allows for disabling dynamic batch weighting.  API slightly changed: `NewMeanMetric` now
    returns a `MeanMetric` struct, not an interface.
  * Added `StreamingMedianMetric`.
* Package `ml/train/optimizers`:
  * Added `RMSProp()` optimizer.
* Package `ml/layers`
  * Added normalizing 1/sqrt(d_k) factor to attention logits in the MultiHeadAttention layer: this will break current
    models using it.
  * Added `RMSNorm` normalizer.
* `gomlx_checkpoints` command-line tool:
  * Added support for multiple models to allow comparing models.
  * Fixed the printing of metrics with tiny values.
* Package `context`:
  * Allow VariableInitializers to use the `context.Context` itself, with its own random initializer.
  * `DefaultInitializer` now creates an initializer. The new default uses He initializer, the same used in PyTorch.
  * Package `initializers`:
    * They now use the `context` random number generator state, which simplifies things. 
    * `ParamInitialSeed` removed, since the RNG is initialized by `Context.RngStateWithSeed()`.
* Fixed some flaky tests.

<hr /><em>This discussion was created from the release <a href='https://github.com/gomlx/gomlx/releases/tag/v0.21.0'>v0.21.0: 2025/07/01 🌞 Summer Edition 🌞</a>.</em>
